File: core/src/main/java/javabot/dao/JavadocClassDao.java
package javabot.dao;

import javabot.javadoc.JavadocApi;
import javabot.javadoc.JavadocClass;
import javabot.javadoc.JavadocClassVisitor;
import javabot.javadoc.JavadocField;
import javabot.javadoc.JavadocMethod;
import javabot.javadoc.criteria.JavadocClassCriteria;
import javabot.javadoc.criteria.JavadocFieldCriteria;
import javabot.javadoc.criteria.JavadocMethodCriteria;
import org.bson.types.ObjectId;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class JavadocClassDao extends BaseDao<JavadocClass> {
    private static final Logger LOG = LoggerFactory.getLogger(JavadocClassDao.class);

    protected JavadocClassDao() {
        super(JavadocClass.class);
    }

    @SuppressWarnings({"unchecked"})
    public List<JavadocClass> getClass(final JavadocApi api, final String name) {
        final String[] strings = JavadocClassVisitor.calculateNameAndPackage(name);
        final String pkgName = strings[0];
        JavadocClassCriteria criteria = new JavadocClassCriteria(ds);
        criteria.upperName().equal(strings[1].toUpperCase());
        if (api != null) {
            criteria.apiId(api.getId());
        }
        if (pkgName != null) {
            criteria.upperPackageName().equal(pkgName.toUpperCase());
        }
        return criteria.query().asList();
    }

    @SuppressWarnings({"unchecked"})
    public JavadocClass[] getClass(final JavadocApi api, final String pkg, final String name) {
        final JavadocClassCriteria criteria;
        try {
            criteria = new JavadocClassCriteria(ds);
            criteria.apiId(api.getId());
            criteria.upperPackageName().equal(pkg.toUpperCase());
            criteria.upperName().equal(name.toUpperCase());
        } catch (NullPointerException e) {
            LOG.debug("pkg = " + pkg);
            LOG.debug("name = " + name);
            throw e;
        }
        return criteria.query().asList().toArray(new JavadocClass[0]);
    }

    @SuppressWarnings({"unchecked"})
    public List<JavadocField> getField(JavadocApi api, final String className, final String fieldName) {
        final JavadocFieldCriteria criteria = new JavadocFieldCriteria(ds);
        List<JavadocClass> classes = getClass(api, className);
        if (!classes.isEmpty()) {
            JavadocClass javadocClass = classes.get(0);
            criteria.javadocClassId(javadocClass.getId());
            criteria.upperName().equal(fieldName.toUpperCase());
            return criteria.query().asList();
        }
        return Collections.emptyList();
    }

    @SuppressWarnings({"unchecked"})
    public List<JavadocMethod> getMethods(JavadocApi api, final String className, final String methodName,
                                          final String signatureTypes) {
        final List<JavadocClass> classes = getClass(api, className);
        List<JavadocClass> list = new ArrayList<>(classes);
        final List<JavadocMethod> methods = new ArrayList<>();

        while (!list.isEmpty()) {
            JavadocClass javadocClass = list.remove(0);
            if (javadocClass != null) {
                methods.addAll(getMethods(methodName, signatureTypes, javadocClass));
                ObjectId superClassId = javadocClass.getSuperClassId();
                if (superClassId != null) {
                    list.add(find(superClassId));
                }
            }
        }
        return methods;
    }

    private List getMethods(final String name, final String signatureTypes, final JavadocClass javadocClass) {
        final JavadocMethodCriteria criteria = new JavadocMethodCriteria(ds);
        criteria.javadocClassId(javadocClass.getId());
        criteria.upperName().equal(name.toUpperCase());
        if (!"*".equals(signatureTypes)) {
            criteria.or(
                           criteria.shortSignatureTypes().equal(signatureTypes),
                           criteria.longSignatureTypes().equal(signatureTypes)
                       );
        }
        return criteria.query().asList();
    }

    public void delete(JavadocClass javadocClass) {
        deleteFields(javadocClass);
        deleteMethods(javadocClass);
        super.delete(javadocClass);
    }

    private void deleteFields(final JavadocClass javadocClass) {
        JavadocFieldCriteria criteria = new JavadocFieldCriteria(ds);
        criteria.javadocClassId(javadocClass.getId());
        ds.delete(criteria.query());
    }

    private void deleteMethods(final JavadocClass javadocClass) {
        JavadocMethodCriteria criteria = new JavadocMethodCriteria(ds);
        criteria.javadocClassId(javadocClass.getId());
        ds.delete(criteria.query());
    }

    public long countMethods() {
        return ds.getCount(JavadocMethod.class);
    }

    public void deleteFor(final JavadocApi api) {
        if (api != null) {
            LOG.debug("Dropping fields from " + api.getName());
            JavadocFieldCriteria criteria = new JavadocFieldCriteria(ds);
            criteria.apiId(api.getId());
            ds.delete(criteria.query());

            LOG.debug("Dropping methods from " + api.getName());
            JavadocMethodCriteria method = new JavadocMethodCriteria(ds);
            method.apiId(api.getId());
            ds.delete(method.query());

            LOG.debug("Dropping classes from " + api.getName());
            JavadocClassCriteria klass = new JavadocClassCriteria(ds);
            klass.apiId(api.getId());
            ds.delete(klass.query());
        }
    }
}

File: src/main/kotlin/javabot/dao/JavadocClassDao.kt
package javabot.dao

import dev.morphia.Datastore
import dev.morphia.query.QueryException
import dev.morphia.query.filters.Filters.eq
import dev.morphia.query.filters.Filters.or
import jakarta.inject.Inject
import java.util.Locale
import javabot.model.javadoc.JavadocApi
import javabot.model.javadoc.JavadocClass
import javabot.model.javadoc.JavadocField
import javabot.model.javadoc.JavadocMethod

class JavadocClassDao @Inject constructor(ds: Datastore) :
    BaseDao<JavadocClass>(ds, JavadocClass::class.java) {
    companion object {
        private fun calculateNameAndPackage(value: String): Pair<String?, String> {
            var clsName = value
            while (clsName.contains(".") && Character.isLowerCase(clsName[0])) {
                clsName = clsName.substring(clsName.indexOf(".") + 1)
            }
            val pkgName =
                if (value != clsName) {
                    value.substring(0, value.indexOf(clsName) - 1)
                } else {
                    null
                }

            return pkgName to clsName
        }
    }

    fun count() = ds.find(JavadocClass::class.java).count()

    fun getClass(api: JavadocApi? = null, name: String): List<JavadocClass> {
        val strings = calculateNameAndPackage(name)
        val pkgName = strings.first
        val query =
            ds.find(JavadocClass::class.java)
                .filter(eq("upperName", strings.second.uppercase(Locale.getDefault())))
        api?.id?.let { id -> query.filter(eq("apiId", id)) }
        if (pkgName != null) {
            query.filter(eq("upperPackageName", pkgName.uppercase(Locale.getDefault())))
        }
        return query.iterator().toList()
    }

    private fun getClassByFqcn(fqcn: String): JavadocClass {
        try {
            return ds.find(JavadocClass::class.java).filter(eq("fqcn", fqcn)).first()
                ?: throw QueryException("Could not find class by fqcn: $fqcn")
        } catch (e: IllegalStateException) {
            throw IllegalStateException("fqcn = ${fqcn}", e)
        }
    }

    fun getClass(api: JavadocApi? = null, pkg: String, name: String): JavadocClass? {
        val query = ds.find(JavadocClass::class.java)
        if (api != null) {
            query.filter(eq("apiId", api.id))
        }
        query
            .filter(eq("upperPackageName", pkg.uppercase(Locale.getDefault())))
            .filter(eq("upperName", name.uppercase(Locale.getDefault())))

        return query.first()
    }

    fun getField(api: JavadocApi?, className: String, fieldName: String): List<JavadocField> {
        val classes = getClass(api, className)
        return if (classes.isNotEmpty()) {
            ds.find(JavadocField::class.java)
                .filter(eq("javadocClass", classes[0]))
                .filter(eq("upperName", fieldName.uppercase(Locale.getDefault())))
                .iterator()
                .toList()
        } else emptyList()
    }

    fun getMethods(
        api: JavadocApi?,
        className: String,
        methodName: String,
        signatureTypes: String
    ): List<JavadocMethod> {
        val classes = getClass(api, className)
        val list = ArrayList(classes)
        val methods = ArrayList<JavadocMethod>()

        list.forEach { klass ->
            methods.addAll(getMethods(methodName, signatureTypes, klass))
            klass.parentTypes.forEach {
                try {
                    methods += getMethods(methodName, signatureTypes, getClassByFqcn(it))
                } catch (e: QueryException) {
                    // parent type isn't in the bot
                }
            }
        }
        return methods
    }

    private fun getMethods(
        name: String,
        signatureTypes: String,
        javadocClass: JavadocClass
    ): List<JavadocMethod> {
        val query =
            ds.find(JavadocMethod::class.java)
                .filter(eq("classId", javadocClass.id!!))
                .filter(eq("upperName", name.uppercase(Locale.getDefault())))
        if ("*" != signatureTypes) {
            query.filter(
                or(
                    eq("shortSignatureTypes", signatureTypes),
                    eq("longSignatureTypes", signatureTypes)
                )
            )
        }
        return query.iterator().toList()
    }

    fun delete(javadocClass: JavadocClass) {
        ds.find(JavadocField::class.java).filter(eq("javadocClass", javadocClass)).delete()

        ds.find(JavadocMethod::class.java).filter(eq("classId", javadocClass.id!!)).delete()

        super.delete(javadocClass)
    }
}
